src/common/config_opts.h

```diff
@@ -899,6 +899,7 @@ OPTION(osd_min_pg_log_entries, OPT_U32, 3000)  // number of entries to keep in t
 OPTION(osd_max_pg_log_entries, OPT_U32, 10000) // max entries, say when degraded, before we trim
 OPTION(osd_force_recovery_pg_log_entries_factor, OPT_FLOAT, 1.3) // max entries factor before force recovery
 OPTION(osd_pg_log_trim_min, OPT_U32, 100)
+OPTION(osd_pg_log_dup_track_versions, OPT_U32, 1000) // how many versions back to track operations that could be duplicated in the pg log
```

Comment:
As implemented this is a trailing number of dups stored past the end of the pg log - I like the idea of (3) though - just making this new option control total dup storage (pg log + dup log), along with decreasing the default pg log length. This is effectively (1) and (3), since there's already dynamic adjustment of how much we trim (whether to the min or max) based on being degraded or in recovery/backfill. See `PrimaryLogPG::calc_trim_to()`.